src/vm_pages.rs

```diff
@@ -1395,18 +1395,23 @@ impl<'a, T: GuestStagePagingMode> FinalizedVmPages<'a, T> {
         let converted = self
             .inner
             .root
-            .get_invalidated_pages(page_addr, num_pages * page_size as u64, |pa, ps| {
+            .get_invalidated_pages(page_addr, num_pages * page_size as u64, |pa, ps, va| {
                 // Converted pages should always be 4k given the host is
                 // exclusively mapped with 4k pages.
                 if ps.is_huge() {
                     return false;
                 }
 
-                // We need to check the first page is suitably aligned and that every page
-                // is contiguous to the previous one. That's because we have some logic below
-                // that allows to create huge pages out of 4k pages, by skipping the intermediate
-                // pages. That's why we must ensure the whole range is physically contiguous and
-                // suitably aligned.
+                // We need to check the first subpage is suitably aligned and that every following
+                // subpage covering the same huge page is contiguous to the previous one. That's
+                // because we have some logic below that allows to create huge pages out of 4k pages,
+                // by skipping the intermediate pages. That's why we must ensure the whole range is
+                // made of physically contiguous subpages for each huge page and they are suitably
+                // aligned.
+                if va.is_aligned(page_size) {
+                    // This is our way of resetting the check for contiguous subpages.
+                    prev_pa = None;
+                }
```

Comment:
I don't think we need to propagate the vaddr to the closure in order to perform this check -- if incrementing `prev_pa` moves us to a `page_size` boundary, just set `prev_pa` to `None`.

Reply:
Done!